File: Godot/WindowPlayer.cs
using Godot;

namespace Game5;

public partial class WindowPlayer : CharacterBody3D, IPlayer
{
	private const float FallAcceleration = 9.8f;
	private IWindowPlayerController _controller;
	private Mesh _mesh;
	private bool _materialized = true;

	public override void _Ready()
	{
		base._Ready();
		// Dematerialize();
		_controller = new WindowPlayerController(this);
	}

	public override void _Process(double delta)
	{
		base._Process(delta);
		_controller.ProcessMovementControls();
	}

	public override void _PhysicsProcess(double delta)
	{
		var velocity = Velocity;
		
		if (_materialized && !IsOnFloor())
		{
			velocity.Y -= FallAcceleration * (float)delta;
		}

		Velocity = velocity;
		
		MoveAndSlide();
	}

	public override void _Input(InputEvent @event)
	{
		base._Input(@event);
		_controller.ProcessMouseControls(@event);
	}

	public void Materialize()
	{
		_materialized = true;
		GetNode<CollisionShape3D>("CollisionShape3D").Disabled = false;
		GetNode<MeshInstance3D>("Pivot/MeshInstance3D").Mesh = _mesh;
		Input.MouseMode = Input.MouseModeEnum.Captured;
	}
	
	public void Dematerialize()
	{
		_materialized = false;
		GetNode<CollisionShape3D>("CollisionShape3D").Disabled = true;
		_mesh = GetNode<MeshInstance3D>("Pivot/MeshInstance3D").Mesh;
		GetNode<MeshInstance3D>("Pivot/MeshInstance3D").Mesh = null;
		Input.MouseMode = Input.MouseModeEnum.Visible;
	}

	public bool IsMaterialized()
	{
		return _materialized;
	}

	public IPlayerController GetController()
	{
		return _controller;
	}
}

File: Sources/Game/WindowPlayer.swift
//
//  WindowPlayer.swift
//  
//
//  Created by Anthony Michalek on 2/18/24.
//

import SwiftGodot;

@Godot
class WindowPlayer: CharacterBody3D, PlayerProtocol {
    private final var fallAcceleration: Float = 9.8;
    private var controller: WindowPlayerControllerProtocol?;
    private var mesh: Mesh?;
    private var materialized: Bool = true;
    
    override func _ready() {
        super._ready();
        // dematerialize();
        controller = WindowPlayerController(windowPlayer: self);
    }
    
    override func _process(delta: Double) {
        super._process(delta: delta);
        controller?.processMovementControls();
    }
    
    override func _physicsProcess(delta: Double) {
        if Engine.isEditorHint() { return }
        if (materialized && isOnFloor())
        {
            velocity.y -= fallAcceleration * Float(delta);
        }
        
        moveAndSlide();
        super._physicsProcess(delta: delta);
        
    }
    
    override func _input(event: InputEvent) {
        super._input(event: event);
        controller?.processMouseControls(event: event);
    }
    
    func materialize() {
        materialized = true;
        (getNode(path: "CollisionShape3D") as! CollisionShape3D).disabled = false;
        (getNode(path: "Pivot/MeshInstance3D") as! MeshInstance3D).mesh = mesh;
        Input.mouseMode = Input.MouseMode.captured;
    }
    
    func dematerialize() {
        materialized = false;
        (getNode(path: "CollisionShape3D") as! CollisionShape3D).disabled = true;
        mesh = (getNode(path: "Pivot/MeshInstance3D") as! MeshInstance3D).mesh;
        (getNode(path: "Pivot/MeshInstance3D") as! MeshInstance3D).mesh = nil;
        Input.mouseMode = Input.MouseMode.visible;
    }
    
    func isMaterialized() -> Bool {
        return materialized;
    }
    
    func getController() -> PlayerControllerProtocol {
        return controller!;
    }
    
}
